File: 322. Coin Change.cpp
class Solution {
public:
    int dp[10001][13];
    int FindMinCoins(vector<int>&coins,int target,int idx){
        if(target<0){
            return 1e9;
        }
        if(idx==coins.size()){
            if(target==0){
                return 0;
            }
            else{
                return 1e9;
            }
        }
        if(dp[target][idx]!=-1){
            return dp[target][idx];
        }
        int include=0,exclude=0;
        include=1+FindMinCoins(coins,target-coins[idx],idx);
        exclude=0+FindMinCoins(coins,target,idx+1);

        return dp[target][idx]=min(include,exclude);
    }
    int coinChange(vector<int>& coins, int amount) {
        memset(dp,-1,sizeof(dp));
        int ans=FindMinCoins(coins,amount,0);
        if(ans==1e9){
            return -1;
        }
        return ans;
    }
};

File: 322. Coin Change.java
class Solution {
    HashMap<String,Integer>dp=new HashMap<>();
    public int minCoins(int[] coins,int amount,int i,int target){
        if(target>amount){
            return 1000000;
        }
        if(i>=coins.length){
            return 1000000;
        }

        if(target==amount){
            return 0;
        }

       String str = Integer.toString(i)+"#"+Integer.toString(target);

       if(dp.containsKey(str)==true){
        return dp.get(str);
       }

        int includeCoins = 1+minCoins(coins,amount,i,target+coins[i]);
        int excludeCoins =   minCoins(coins,amount,i+1,target);

        int tempAns=Math.min(includeCoins,excludeCoins);
        dp.put(str,tempAns);

        return tempAns;
    }
    public int coinChange(int[] coins, int amount) {
        int n=coins.length;
        int ans = minCoins(coins,amount,0,0);
        if(ans==1000000){
            return -1;
        }
        return ans;
    }
}
